src/toplevel.c

```diff
@@ -131,48 +147,59 @@ static jl_value_t *jl_eval_module_expr(jl_module_t *parent_module, jl_expr_t *ex
         jl_type_error("module", (jl_value_t*)jl_symbol_type, (jl_value_t*)name);
     }
 
-    jl_module_t *newm = jl_new_module(name);
-    jl_value_t *form = (jl_value_t*)newm;
-    JL_GC_PUSH1(&form);
-    JL_LOCK(&jl_modules_mutex);
-    ptrhash_put(&jl_current_modules, (void*)newm, (void*)((uintptr_t)HT_NOTFOUND + 1));
-    JL_UNLOCK(&jl_modules_mutex);
-
-    // copy parent environment info into submodule
-    newm->uuid = parent_module->uuid;
-    if (jl_is__toplevel__mod(parent_module)) {
-        newm->parent = newm;
-        jl_register_root_module(newm);
-    }
-    else {
+    jl_binding_t *b = jl_get_binding_wr(parent_module, name, 1);
+    jl_module_t *newm;
+    jl_value_t *old = b->value;
+    if (old == NULL) {
+        if (parent_module == jl_main_module) // jl_check_open_for allowed this, but now we must block it
+            if (jl_options.incremental && jl_generating_output())
+                jl_errorf("Evaluation into the closed module `Main` breaks incremental compilation",
+                          "because the side effects will not be permanent.");
+        newm = jl_new_module(name);
+        // copy parent environment info into submodule
+        newm->uuid = parent_module->uuid;
         newm->parent = parent_module;
-        jl_binding_t *b = jl_get_binding_wr(parent_module, name, 1);
         jl_declare_constant(b);
-        jl_value_t *old = NULL;
-        if (!jl_atomic_cmpswap(&b->value, &old, (jl_value_t*)newm)) {
-            if (!jl_is_module(old)) {
-                jl_errorf("invalid redefinition of constant %s", jl_symbol_name(name));
-            }
-            if (jl_generating_output())
-                jl_errorf("cannot replace module %s during compilation", jl_symbol_name(name));
-            jl_printf(JL_STDERR, "WARNING: replacing module %s.\n", jl_symbol_name(name));
```

Comment:
This will break some workflows.

Reply:
workflows that print WARNINGs are not stable